### DISCUSSION TITLE
Decoupling the component of AS4, SMP-Client and Others in Spring Boot Project

### DISCUSSION THREAD

**Discussion by Upendra0** [phax/phase4#240] [Q&A]
Hi **@phax**,
I am using the spring boot demo project to send and receive document using AS4.
The project repository link is: https://github.com/phax/phase4/tree/master/phase4-spring-boot-demo

The builder class which is responsible to send the AS4 Message seems to be not have only one responsibility. It is not only taking an AS4 Message Payload and sending it. 
It is doing other task also, for e.g.:

1.  Getting the receiver details using SMP
2.  Adding the Standard Business Document Header in the final document which needs to be sent.

This violates the Single Responsibility Principle and makes difficult to manage or to change the project. 
Is there a way such that i can Separate these components? 
Is there a component which i can use only to send AS4 message and that component should not do anything else apart from sending the message. All the details which this component need, I will provide to it. This will make the component loosely coupled.

**Comment by phax** [comment]
Hello @Upendra0 
Thanks for taking the time to write this discussion item.
Indeed, `phase4-spring-boot-demo` is tailored towards the Peppol Network and is not an indepedent AS4 solution.

However, the builder classes are imho nicely strctured. For the Peppol client, the submodule `phase4-peppol-client` contains 2 specific builder classes `Phase4PeppolSender.Builder` and `Phase4PeppolSender.SBDHBuilder` that both inherit from `Phase4PeppolSender.AbstractPeppolUserMessageBuilder`. These builder classes all add the Peppol details onto the more generic builder class `AbstractAS4UserMessageBuilderMIMEPayload` (in module `phase4-lib`).

If you already have the SBDH built yourself, you choose the `Phase4PeppolSender.SBDHBuilder` whereas if you want the SBDH to be created for you, you choose the `Phase4PeppolSender.Builder`. For me that is a good separation of concern.

If you want the SMP lookup done for you, pass in the `ISMPServiceMetadataProvider` interface. If you've done the SMP lookup before yourself (or used some kind of caching), then you can pass in the the specific values via `receiverEndpointDetails(X509Certificate, String)` in the builder class.

So basically there are different ways to fill the data needed for sending so that you have the freedom of choice to do whatever you need to do. If you are missing something in particular, I am happy to add it.
